### PR TITLE
Added support for the musicbrainz "ARTISTS" tag in id3v2, ape, xiph.

### PR DESCRIPTION
**Pull request by evilhamster**
As subject, it addes support for the musicbrainz ARTISTS tag into kodi. This solves the problem reported in http://forum.kodi.tv/showthread.php?tid=198149 for musicfiles that have this tag set correctly. 

I have tested this change against files that have tags written with picard and/or puddletag, so very minor testing has been done.

### CODE REVIEW

**Comment by zag2me** [comment]
Hurray!

Will need more testing though.

**Comment by MartijnKaijser** [comment]
@mkortstiege is this something that can be considered bugfix or defer to 16.0?

**Comment by mkortstiege** [comment]
Will this produce dupe artists in case `ARTIST` and `ARTISTS` contain the same or overlapping meta? As the musiclibrary is not really my department and since I am not sure whether or not this might introduce dupes and/or regressions i tend to defer to 16.0.

Would like to hear what @Montellese thinks about this.

**Comment by evilhamster** [comment]
It should not produce any dupe artists in that case, if the artists tag is set it will use that one instead of the artist one. For both the artist and artists tags it calls the setartist function and that one overwrites what has previously been set.

**Comment by mkortstiege** [comment]
Right. Misread that diff, sorry. No objections for 15.0 then, but i would like to get another +1.
So should be good to go after another brief review and @zag2me`s feedback.

**Comment by MartijnKaijser** [comment]
Creating test builds now. All my music is MusicBrainz tagged so I could do a run and provide the database for analysing

**Comment by MartijnKaijser** [comment]
@zag2me builds are available in the server with hash 4b02c0f
Here's my db for analysis after fresh scan.
https://dl.dropboxusercontent.com/u/7435592/MyMusic52.db

@mkortstiege (or some one else)
i just created a fresh scan. after switching on "library" in the left panel i get asked to scan again. Could you look into killing that? iirc it was from Frodo days. It's some upgrade flag from when @night199uk redesigned the database and it's been hunting us ever since.

File: xbmc/music/tags/TagLoaderTagLib.cpp [hunk 1]
```diff
@@ -448,6 +448,8 @@ bool CTagLoaderTagLib::ParseID3v2Tag(ID3v2::Tag *id3v2, EmbeddedArt *art, CMusic
           SetAlbumArtist(tag, StringListToVectorString(stringList));
         else if (desc == "ALBUM ARTIST")
           SetAlbumArtist(tag, StringListToVectorString(stringList));
+        else if (desc == "ARTISTS") 
+          SetArtist(tag,StringListToVectorString(TagLib::StringList::split(stringList.front(), TagLib::String("/"))));
```
Comment (Montellese) [inline xbmc/music/tags/TagLoaderTagLib.cpp]: Is `/` a fixed character defined by the ID3v2 standard?

**Comment by Montellese** [comment]
What I don't like is that the preference of either `ARTIST` or `ARTISTS` is not explicit at all in the code and heavily depends on the ordering of the iterators of `APE::ItemListMap` and `Ogg::FieldListMap`. So this is a completely implicit and depends on the implementation of those maps.

Apart from that I'm no expert on music tags or our music library/database and I don't have any MusicBrainz tagged music so can't really test it.

**Comment by evilhamster** [comment]
@Montellese very good comments. 

Assuming the order of the artists and artists field in the maps might not have been a good idea, I will update the pr to handle it.

From my understanding / is a standard in id3v2.3 but not id3v2.4. All my test cases used 2.3 so I did not see this problem. I will update this pr to handle it. 

This pr should not be merged as it currently stands.

**Comment by evilhamster** [comment]
Updated the pr so that it supports both id3 2.3 and 2.4. Also updated the code in accordance to Montellese comment.

**Comment by zag2me** [comment]
During testing of this, I noticed Picard can also be set to separate artists with ; and with whitespace/ on id3v2.3 tags.
I doubt its used very much but should we respect it?

image3

**Comment by zag2me** [comment]
Before PR
image2

After PR
image3

Nice work!
(this was using the test build posted by @MartijnKaijser)

I did notice I have to use a recent MusicBrainz Picard (v1.3.2) version that supports the "artists" tag.

**Comment by zag2me** [comment]
Spoke too soon, on further testing I'm getting duplicate artists.

Test files here

http://cloud.mgdn.nl/index.php/s/f0HYGYK3vgP4mQW?path=%2F

**Comment by evilhamster** [comment]
That test build was made yesterday, the fix I made to handle id3 2.4 was made 2 hours ago :)

I tested it with the latest version (in this pr) and the files you listed and they work fine.

**Comment by zag2me** [comment]
excellent, if someone can make another test build, i'll get some others to test as well.

**Comment by MartijnKaijser** [comment]
building new one which hash 68a0a6e (available in about an hour. same place)

**Comment by Rouzax** [comment]
The test I've done are very promising. 
Because it is based on Kodi 15 and I still run 14 on my live system I did some testing with a new installation and 2 albums, one compilation and one regular artist album.

Under the Artist page it looks very nice.
image

The songs under the album are a lot better than they were but it appears that all the join phrase like: "feat." or "vs." are disregarded. 
image

If it is possible to implement that it would be a nice to have but if that is to difficult to add to Kodi 15 than please add this to begin with.

All in all, thank you for this :+1:

**Comment by meowmoo** [comment]
id3 2.4 mp3's looks good
[URL=http://s843.photobucket.com/user/emteeelp/media/screenshot004.png.html][IMG]http://i843.photobucket.com/albums/zz360/emteeelp/screenshot004.png[/IMG][/URL]

**Comment by Rouzax** [comment]
Forgot to tell, my mp3's are id3 2.4

File: xbmc/music/tags/TagLoaderTagLib.cpp [hunk 2]
```diff
@@ -448,6 +450,17 @@ bool CTagLoaderTagLib::ParseID3v2Tag(ID3v2::Tag *id3v2, EmbeddedArt *art, CMusic
           SetAlbumArtist(tag, StringListToVectorString(stringList));
         else if (desc == "ALBUM ARTIST")
           SetAlbumArtist(tag, StringListToVectorString(stringList));
+        else if (desc == "ARTISTS" && id3v2->header()->majorVersion() == 3)
```
Comment (Montellese) [inline xbmc/music/tags/TagLoaderTagLib.cpp]: I would change this to
```
else if (desc == "ARTISTS")
{
  artistsTagFound = true;
  if (id3v2->header()->majorVersion() < 4)
    SetArtist(tag,StringListToVectorString(TagLib::StringList::split(stringList.front(), TagLib::String("/"))));
  else
    SetArtist(tag,StringListToVectorString(stringList));
}
```
which saves a few lines and is a bit more generic.

**Comment by Montellese** [comment]
Thanks for the adjustments. Another thought that came to mind since I don't know how ID3 tags work in detail: Is it possible that a tag like `ARTISTS` is present but empty? That case would require additional checks to only set `artistsTagFound` if the `ARTISTS` tag isn't empty. If that's not a possibility than everything is ok from a coding point of view. Can't really comment on whether this makes sense and whether people prefer to have `Foo feat. Bar` as an artist or two separate artist entries `Foo` and `Bar`.

**Comment by zag2me** [comment]
Just confirming this works for me now, with both 2.3 and 2.4 tags.

No more duplicates.

**Comment by scott967** [comment]
The concern I have about this fix is this:  The original problem of "duplicate artists" appears to be a result (but not necessarily only) of users who use Musicbrainz Piccard tagging software.  This tagger in turn uses conventions from the Musicbrainz database v2 for the semantics.  These users typically have stored in the "artist" tag what Musicbrainz defines as an "artist credit".  This often is created as a text string with various forms of connective.  For example: "Sonny Til & The Orioles", or "Hardwell Feat. Matthew Koma".  Musicbrainz deals with the ambiguity of the "artist credit" by using separate tags for the "artist" and "artist alias" (if any) which is tagged as "artists" and Musicbrainz artist ID.  I'm not aware if Musicbrainz requires every artist in the "artists" tag to have an artist ID, but it makes sense that they would.  Since the ID is essentially a GUID it greatly improves searching compared to using text strings.

My understanding is that users who follow the Musicbrainz Piccard tagging format would like to have the "artist credit" stored in the database and displayed as an identifier for individual song files.  Currently Kodi does have the concept of "song artist" in the music database which provides a link to both the artist table and the text string strArtist.  So as a result the "artist credit"/song artist must also reside in the artist table.  

What this PR does is essentially replace the "artist credit" / song artist with the contents of the "artists" tag.  This keeps the song artist and artist tables consistent, but loses the concept of "artist credit".  The same could be accomplished without the PR if users simply use the "artist" tag as intended by specification rather than as modified by Musicbrainz.  I have no opinion on that since I don't use Piccard tagging software.  If Piccard users are happy with this result then it fixes the immediate problem, but I thought the whole idea of the Musicbrainz "artist credit" was to retain that data and I assume that is something Piccard users are interested in.

**Comment by evilhamster** [comment]
@Montellese 
The code changes you suggested was way cleaner and has been adapted. I'm not sure if a TXXX fram can exist that has zero length and have been unable to create one using picard (the only software I know that uses this tag), so I don't think we need to handle that case. Let me know if you think otherwise and I'll update it.

@scott967 
The problem isn't specifically with picard but using the mbrainz_artist_id as identifiers for the artist (instead of using the string value from the artist field).

So for example if you have a song:
TPE1 (Lead performer(s)/Soloist(s)): Hardwell feat. Matthew Koma
TXXX (User defined text information): (MusicBrainz Artist Id): 170d6637-36e0-49dd-a225-c4fd5b77a285;f387841c-9bc7-47e9-b78b-6fb7834ac71f

Kodi will try and add two artists to the database, but since there is no information on the names of the different artists they will both be added to the database with the name "Hardwell feat. Matthew Koma" where the end result would be two artists with the same name (unless you used online sources to update the names, but this would still cause issues since it would update the name of the artists in the db but not the information in the song titles so they would still be duplicates). To sort this problem you either need to ignore the mbrainz tags or do something like this pr.

**Comment by Rouzax** [comment]
In the ideal world I would have to agree with @scott967. The PR as is removes all join phrases from the song information. However as a first step this is already way better than what we have now.

**Comment by Maaarv7** [comment]
I think that a problem with this solution is when you have stored the "feat."-information in the title rather than in the artist field.

For example you have:
Artist: Eminem
Album Artist: Eminem
Title: Monster (feat. Rihanna)
Artists: Eminem / Rihanna
Music_Brainz_Artist_ID: XXXX-XXXX / YYYY-YYYY

As it is right now you suffer from the reported problem with multiple artist entries when you have the musicbrainz_artist_ids set. So of course the library will show:

Eminem / Eminem - Monster (feat. Rihanna)

This PR solves this and is very welcomed of course:
But now you have kind of a double feat.-information in the music library which does not look nice:

Eminem / Rihanna - Monster (feat. Rihanna)

I'm no expert in this but isn't it possible to get the standard artist tag shown while displaying single songs? This would also solve the problem when you have stored something like "&" or "vs." in the artist field. AND you would be able to browse for Rihanna as single artist.

**Comment by Rouzax** [comment]
MusicBrainz does not allow featuring artists in the title. They should always be placed with the artists. 
So tag you have should not exist.

**Comment by scott967** [comment]
I understand that the problem was triggered by the artist MBID, specifically the case that the number of scanned artist MBIDs is greater than the number of scanned artists.  But the problem is that the code currently expects that the artist tag contain the same or greater number of individual artists than the artist MBID tag (based on my testing).  There also appears to be an implicit ordering of the two tags, tough I guess the "prefer online information" switch can mitigate this (I'm not sure that was intended but rather a side-effect).  In my way of thinking, this is more of a band-aide solution which works in this case, but doesn't address the larger issue of the intended meaning of the artist tag and how "prefer online information" should work in regards to adding new artists into the database.

**Comment by zag2me** [comment]
I just had a quick chat with the musicbrainz developers and they pointed towards this standard:

https://musicbrainz.org/doc/Style/Artist_Credits/Featured_artists

They didn't see anyway of fixing the problem easily, but scanning the artist credit tag for "feat." and only adding the artist tag if this is present may work.

Basically my opionion is this Pull Request should be merged now for 15 as it fixes the duplicate artist bug, and we can work on feat. artists for 16. @MartijnKaijser let me know what you think?

NOTE: Also for reference in the future, there is a "Primary Artist" tag coming in picard hopefully at some point https://github.com/musicbrainz/picard/pull/369

**Comment by GitrianDK** [comment]
MusicBrainz Picard has a plugin which moves the featuring artist to the title, so this is indeed a possible situation.

 I think there is 2 different cases. First there is the organising, so that a song with several artist is indexed and displayed under all artists. Secondly there is the display of the song which should be an exact copy of what is in the artist & title tags.

**Comment by Maaarv7** [comment]
What GitrianDK says, if this can be merged for v16 maybe, this would be top notch.

**Comment by night199uk** [comment]
Indexing is different from the display artists string, as people have already mentioned. There are many cases here to handle, and Musicbrainz does that pretty well.

The idea behind the musicbrainz artists tag when I PRd it to Picard is that it should be a 1:1 mapping with the MusicBrainz Artist IDs, and it should include the 'primary' name of the artist. This way, we can seed the database with a reasonable Artist ID -> Artist Name mapping even if we run offline. I hadn't added support for it as it wasn't there, and I had to PR it to Picard. BTW, Jaikoz already wrote this tag.

The 'display' string for a song is different, and can be built by the scraper using XSLT IIRC.

I forget what exactly was in XBMC now, I'll try and take a look over the weekend at this PR, but the ideal is this:
Artist - (a pure display string version of the song artists. For ID3v1 and IDv2.3 with multiple entries should be concatenated with a ' /  ' separator: e.g. 'The Artist Formerly Known as Prince feat. 2Pac')
MusicBrainz Arist IDs -> a ';' separated list of artist IDs (or multiple fields in ID3v2.4)
Artists -> a mapping of the original names for each artist ID in the above. e.g. 'Prince', and '2Pac'.

And the same for Album Artist[s]/IDs. So in the above case, all tracks by 'Prince' would be cataloged under 'Artists' view as 'Prince', but individual song credits could contain 'The Artist Formerly Known as Prince'; that wouldn't show up in the Artist Index but would show up when you play the song.

Now the 'artists' tag has propagated widely - this should be much easier, but I've not looked at the implementation in a while.

**Comment by evilhamster** [comment]
The point of this PR was mostly to make sure that there was a mapping between the artists and musicbrainzid:n to avoid the duplicate the artist issue but try to minimize the amount of code that could be affected. 

I have also been working on some code that will keep the artists tag for display purposes (since @Rouzax mentioned that would be prefered), I currently have some code that seems to be working and uses the original song artist tag info for the display but I need to fix it the same way for the album artists. If you are interested in checking it out, https://github.com/evilhamster/xbmc/tree/addartistdescfield

**Comment by scott967** [comment]
Note that currently in Picard, there is an "artists" tag which is documented as the "Track Artist Name(s) (multi-value) (since Picard 1.3)".  There is no analog for the release (album) artist name(s).  As far as trying to use a release artist(s) credit, ISTM you have to be careful about the relationship of release (album) artist credit and "various artists" / compilation setting or determination.

**Comment by evilhamster** [comment]
http://tickets.musicbrainz.org/browse/PICARD-700 so there is hope we might get a corresponding tag for the albumartists in the future :) 

I did a kinda simplistic version seems to be working fine, but it would be great if someone with more knowledge about the musictag part of the kodi codebase could check my work. 

Do you want me to update this PR with the contents of the branch I posted previously, although it's more complex than this one and there is a bigger chance I missed something or should I wait with that one until 15 is out and work on 16 is underway?

**Comment by zag2me** [comment]
Since we are so late in the release cycle, I personally think this fixes an issue and should be merged as is. Then work on the other display stuff for Version 16. @night199uk @MartijnKaijser?

**Comment by MartijnKaijser** [comment]
i'd like the opinions of @Montellese and @mkortstiege who has some overview as well for musicdb
Music library isn't a highly used feature but when it screw up it screws up good.

**Comment by mkortstiege** [comment]
Too late in the cycle, so v16 if it's up to me.

**Comment by MartijnKaijser** [comment]
i kinda share that opinion. And we could possibly always backport it to a point release when it "proves" itself.

**Comment by Montellese** [comment]
I'm probably not the right person to make the decision because I use the musicdb very rarely and I don't use MusicBrainz tags so I don't even really know what the real issue here is. All I can see is that the fix also seems to come with a drawback (i.e. loosing the `feat.` stuff in the full artist string that we display.

**Comment by scott967** [comment]
I think the point is that if a user chooses to use the new "artists" tag, it is with the understanding he will lose the contents of the current "artist"  tag.  So it is a user decision which way to go.

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by steve1977** [comment]
Fantastic to see the development here. I am one of the folks desperately waiting for a fix.

I can well live with the decision to defer to 16. Ultimately, I will rather early move to 16 and alternatively, there remains the backport option suggested bty Martijn.

Would be great if there could be a way to set "artist" as the displayname and use "artists" for the database. If someone looks into this, would be great to make the same edits for albumartist/albumartists.

Some of you may have noticed that there is a bug in Picard that prevents artists tag to be written for mp4/m4a. For those interested in a fix, maybe you can + in this this thread (http://tickets.musicbrainz.org/browse/PICARD-376). Thanks!

**Comment by steve1977** [comment]
Fantastic to see the development here. I am one of the folks desperately waiting for a fix.

I can well live with the decision to defer to 16. Ultimately, I will rather early move to 16 and alternatively, there remains the backport option suggested bty Martijn.

Would be great if there could be a way to set "artist" as the displayname and use "artists" for the database. If someone looks into this, would be great to make the same edits for albumartist/albumartists.

Some of you may have noticed that there is a bug in Picard that prevents artists tag to be written for mp4/m4a. For those interested in a fix, maybe you can + in this this thread (http://tickets.musicbrainz.org/browse/PICARD-376). Thanks!

**Comment by steve1977** [comment]
Fantastic to see the development here. I am one of the folks desperately waiting for a fix.

I can well live with the decision to defer to 16. Ultimately, I will rather early move to 16 and alternatively, there remains the backport option suggested bty Martijn.

Would be great if there could be a way to set "artist" as the displayname and use "artists" for the database. If someone looks into this, would be great to make the same edits for albumartist/albumartists.

Some of you may have noticed that there is a bug in Picard that prevents artists tag to be written for mp4/m4a. For those interested in a fix, maybe you can + in this this thread (http://tickets.musicbrainz.org/browse/PICARD-376). Thanks!

**Comment by steve1977** [comment]
Fantastic to see the development here. I am one of the folks desperately waiting for a fix.

I can well live with the decision to defer to 16. Ultimately, I will rather early move to 16 and alternatively, there remains the backport option suggested bty Martijn.

Would be great if there could be a way to set "artist" as the displayname and use "artists" for the database. If someone looks into this, would be great to make the same edits for albumartist/albumartists.

Some of you may have noticed that there is a bug in Picard that prevents artists tag to be written for mp4/m4a. For those interested in a fix, maybe you can + in this this thread (http://tickets.musicbrainz.org/browse/PICARD-376). Thanks!

**Comment by steve1977** [comment]
Fantastic to see the development here. I am one of the folks desperately waiting for a fix.

I can well live with the decision to defer to 16. Ultimately, I will rather early move to 16 and alternatively, there remains the backport option suggested bty Martijn.

Would be great if there could be a way to set "artist" as the displayname and use "artists" for the database. If someone looks into this, would be great to make the same edits for albumartist/albumartists.

Some of you may have noticed that there is a bug in Picard that prevents artists tag to be written for mp4/m4a. For those interested in a fix, maybe you can + in this this thread (http://tickets.musicbrainz.org/browse/PICARD-376). Thanks!

**Comment by MartijnKaijser** [comment]
@steve1977 stop spamming

**Comment by steve1977** [comment]
Fantastic to see the development here. I am one of the folks desperately waiting for a fix.

I can well live with the decision to defer to 16. Ultimately, I will rather early move to 16 and alternatively, there remains the backport option suggested bty Martijn.

Would be great if there could be a way to set "artist" as the displayname and use "artists" for the database. If someone looks into this, would be great to make the same edits for albumartist/albumartists.

Some of you may have noticed that there is a bug in Picard that prevents artists tag to be written for mp4/m4a. For those interested in a fix, maybe you can + in this this thread (http://tickets.musicbrainz.org/browse/PICARD-376). Thanks!

**Comment by steve1977** [comment]
Fantastic to see the development here. I am one of the folks desperately waiting for a fix.

I can well live with the decision to defer to 16. Ultimately, I will rather early move to 16 and alternatively, there remains the backport option suggested bty Martijn.

Would be great if there could be a way to set "artist" as the displayname and use "artists" for the database. If someone looks into this, would be great to make the same edits for albumartist/albumartists.

Some of you may have noticed that there is a bug in Picard that prevents artists tag to be written for mp4/m4a. For those interested in a fix, maybe you can + in this this thread (http://tickets.musicbrainz.org/browse/PICARD-376). Thanks!

**Comment by steve1977** [comment]
Fantastic to see the development here. I am one of the folks desperately waiting for a fix.

I can well live with the decision to defer to 16. Ultimately, I will rather early move to 16 and alternatively, there remains the backport option suggested bty Martijn.

Would be great if there could be a way to set "artist" as the displayname and use "artists" for the database. If someone looks into this, would be great to make the same edits for albumartist/albumartists.

Some of you may have noticed that there is a bug in Picard that prevents artists tag to be written for mp4/m4a. For those interested in a fix, maybe you can + in this this thread (http://tickets.musicbrainz.org/browse/PICARD-376). Thanks!

**Comment by zag2me** [comment]
Long shot I know... but is there any chance a backport would be accepted for isengaard? 

This fixes duplicate artists if your music is tagged with the ARTISTS tag.

**Comment by steve1977** [comment]
Curious what is spam about my post above...

**Comment by MartijnKaijser** [comment]
@steve1977 somehow you posted the same question about 15 times

**Comment by steve1977** [comment]
Does anyone has the chance to look into a backport? Thanks in advance!

**Comment by hudokkow** [comment]
>Curious what is spam about my post above...

And yet you continue to spam! Every... **single**... member... Every time you or anybody else posts here **we** get a ping/message/mail/whatever.

**Not** development related == spam!

Stop it, please!

**Comment by rjsachse** [comment]
So at the end of day it looks like kodi needs a standed tagging system.

To sove the double artist on (14) & (15) I used picard with the plugin to move feat to title and copy artists to artist so a tag looks like this.

Title = The Festival Song (Feat. 360, Hailey Cramer & Seth Sentry)
Album Artist = Pez
Artist = Pez / 360 / Hailey / Seth Sentry
Artists = Pez / 360 / Hailey / Seth Sentry

Now with that tagging system everything works and looks great no double ups of artist or artist feat. 

So what I am saying is there is a so called fix if you tag your files this way.

I have never had a perfect looking music database till kodi with picard and somr plugin's and scripts, yer not noob friendly.  So my suggestion is to stick to a format and get people to tag files accordingly.
I think kodi should use picard to write the metadata and users who want a clean looking music database just can scan the music hit save and done so there tag looks like this.

Title = The Festival Song
Album Artist = Pez
Artist = Pez feat. 360, Hailey Cramer & Seth Sentry
Artists = Pez / 360 / Hailey Cramer / Seth Sentry

So is there a need for kodi to read artist tag? Unless id3v1
Should music just be scaned like movies using misicbrainz db and a nfo file.

Oh and don't forget if you have a artist feat artist then what about skinners and artist logos.

**Comment by Rouzax** [comment]
Did a test with Jarvis Beta 4 to see if this is resolved and like it says in the release notes it is fixed!
Image

The only "odd" thing is that Various Artists are now an Artist, not sure if that is as intended.
Image

**Pull request merged**